### PR TITLE
make judging from features more convenient, this method expects uses …

### PR DESCRIPTION
**Pull request by russelljjarvis**
This method expects users to have pre-existing code that performs something like judge generate prediction. Prediction is then just treated as a settable object attribute, which can be updated.

Doing things this way is more amenable to high throughput feature generation. Ie if EFEL generates 100 new features per model.

### CODE REVIEW

File: sciunit/tests.py [hunk 1]
```diff
@@ -116,6 +116,10 @@ def validate_observation(self, observation: dict) -> dict:
                           'type': 'dict'}
             schema = {'observation': schema}
             v = ObservationValidator(schema, test=self)
+            if float(observation['std']) == 0.0:
```
Comment (rgerkin) [inline sciunit/tests.py:119]: If you want to add this sort of validation step, you can add methods to ObservationValidator like [this one](https://github.com/scidash/sciunit/blob/master/sciunit/validators.py#L57) and [this one](https://github.com/scidash/sciunit/blob/master/sciunit/validators.py#L69), then those will automatically get run in the `v.validate` line that was already here.  You can call it `_validate_std` and have it ensure that the `std` is not 0.  However, you may want to have this only be a part of validation if the test's `score_type` is `ZScore`.  The test is accessible inside those validation methods as `self.test`.

File: sciunit/tests.py [hunk 2]
```diff
@@ -369,6 +373,97 @@ def _judge(self, model: Model, skip_incapable: bool=True) -> Score:
         # 6.
         self._bind_score(score, model, self.observation, prediction)
 
+
```
Comment (rgerkin) [inline sciunit/tests.py:376]: There is a lot of code duplication here.  We can still just keep `judge` and `_judge`, and add a keyword argument to handle whatever the `presupplied_feature` code is trying to do inside one or both of them.  I think you are just trying to cut out the step where the model gets run.  You could just add `prediction` itself as a keyword argument to `judge` (defaulting to `None`) and then inside `judge` it checks if `prediction` is not `None` and if so it skips the model running code.  This could be as simple as `prediction = prediction if prediction is not None else self.generate_prediction(model)` which will leave prediction alone if has been set, and otherwise it will get it from the model.

**Review by rgerkin — CHANGES_REQUESTED** [review]
See my comments about how this should be implemented.

File: sciunit/tests.py (hunk 1 above)
Comment (russelljjarvis) [reply sciunit/tests.py:119]: inside sciunit.sciunit.validators.ObservationValidator class:

I have done this.
```python 
    def _validate_not_zero(self, has_units: bool, key: str, value: Any) -> None:

        # import done here only because its not expected ZScore will be used in the module outside of this method scope
        from sciunit.scores import ZScore
        if self.test.score_type is ZScore:
            if float(self.test.observation['std']) == 0.0:
                raise ObservationError(v.errors)
```

I have to confess I have no idea what the method signature means I just copied other similar method signatures.

Comment (russelljjarvis) [reply sciunit/tests.py:119]: Applying above code results in this warning.

/home/user/anaconda3/lib/python3.7/site-packages/cerberus/validator.py:1609: UserWarning: No validation schema is defined for the arguments of rule 'not_zero_obs_zscore'

File: sciunit/tests.py (hunk 2 above)
Comment (russelljjarvis) [reply sciunit/tests.py:376]: Okay you are right, I got the same effect by changing the function signatures of `def judge/judge_` to include 

`prediction: dict: None`, etc

Comment (rgerkin) [reply sciunit/tests.py:376]: @russelljjarvis You merged this in but there is still code duplication.  I don't think we need `_presupplied_feature_judge` or `feature_judge` at all because this can just be a kwarg to the actual `_judge` and `judge`.  Can you implement without these extra functions (just put the unique logic into an if statement in the actual `_judge` and `judge`)?

Comment (russelljjarvis) [reply sciunit/tests.py:376]: Its true I don't need
```
_presupplied_feature_judge or feature_judge
```
 I thought I deleted both from the pull request.

let me try again.

Comment (russelljjarvis) [reply sciunit/tests.py:376]: Also I accidentally pushed directly to the branch, sorry about that. I since deleted `git remote rm origin`, so I can't do that by accident anymore. The commits might now be aligned with your requests if you can see through the confusion.

**Comment by russelljjarvis** [comment]
Still having trouble with using prediction=None in the method signature, I think this problem is less straight forward than it seems.

**Comment by rgerkin** [comment]
What is the specific problem?  Inside the old code, it generates the local variable `prediction` by setting it equal to the result of `generate_prediction`.  So you should be able to pass one in (with default value `None` and only call `generate_prediction` if it is still `None`).

**Comment by russelljjarvis** [comment]
I put a break point in the code, I think there were two problems 

# 1
`_judge`, or `judge` was not returning anything (esp score), perhaps because of me clumsily deleting the return statement in the editor (fixed now).

#2 I had not nested the statement validate model outside of condition `if prediction is None:`
fixed now. 

I will push code and update PR.

**Comment by russelljjarvis** [comment]
See
https://github.com/scidash/sciunit/pull/174/commits/829145abd6aa67dd915d356b00de442df9ad2efc
